worker/secretbackendrotate/rotate.go

```diff
@@ -0,0 +1,229 @@
+	if config.SecretBackendManagerFacade == nil {
+		return errors.NotValidf("nil Facade")
+	}
+	if config.Clock == nil {
+		return errors.NotValidf("nil Clock")
+	}
+	if config.Logger == nil {
+		return errors.NotValidf("nil Logger")
+	}
+	return nil
+}
+
+// NewWorker returns a Secret Backend token rotation Worker backed by config, or an error.
+func NewWorker(config Config) (worker.Worker, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	w := &Worker{
+		config:      config,
+		backendInfo: make(map[string]tokenRotateInfo),
+	}
+	err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+	})
+	return w, errors.Trace(err)
+}
+
+type tokenRotateInfo struct {
+	ID          string
+	backendName string
+	rotateTime  time.Time
+	whenFunc    func(rotateTime time.Time) time.Duration
+}
+
+func (s tokenRotateInfo) GoString() string {
+	return fmt.Sprintf("%s token rotation: in %v at %s", s.backendName, s.whenFunc(s.rotateTime), s.rotateTime.Format(time.RFC3339))
+}
+
+// Worker fires events when secrets should be rotated.
+type Worker struct {
+	catacomb catacomb.Catacomb
+	config   Config
+
+	backendInfo map[string]tokenRotateInfo
+
+	timer       clock.Timer
+	nextTrigger time.Time
+}
+
+// Kill is defined on worker.Worker.
+func (w *Worker) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+// Wait is part of the worker.Worker interface.
+func (w *Worker) Wait() error {
+	return w.catacomb.Wait()
+}
+
+func (w *Worker) loop() (err error) {
+	changes, err := w.config.SecretBackendManagerFacade.WatchTokenRotationChanges()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err := w.catacomb.Add(changes); err != nil {
+		return errors.Trace(err)
+	}
+	for {
+		var timeout <-chan time.Time
+		if w.timer != nil {
+			timeout = w.timer.Chan()
+		}
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case ch, ok := <-changes.Changes():
+			if !ok {
+				return errors.New("secret rotation change channel closed")
+			}
+			w.handleTokenRotateChanges(ch)
+		case now := <-timeout:
+			if err := w.rotate(now); err != nil {
+				return errors.Annotatef(err, "rotating secret backends")
+			}
+		}
+	}
+}
+
+func (w *Worker) rotate(now time.Time) error {
+	w.config.Logger.Debugf("processing secret backend token rotation at %s", now)
+
+	var toRotate []string
+	for id, info := range w.backendInfo {
+		w.config.Logger.Debugf("checking %s: rotate at %s... time diff %s", id, info.rotateTime, info.rotateTime.Sub(now))
+		// A one minute granularity is acceptable for secret rotation.
+		if info.rotateTime.Truncate(time.Minute).Before(now) {
+			w.config.Logger.Debugf("rotating token for %s", info.backendName)
+			toRotate = append(toRotate, id)
+			// Once backend has been queued for rotation, delete it here since
+			// it will re-appear via the watcher after the rotation is actually
+			// performed and the last rotated time is updated.
+			delete(w.backendInfo, id)
```

Comment:
It might be still ok to delete them here even if the later RotateBackendTokens() call failed.
Because the watcher will notify us again later, right?

Reply:
Yes, worker will restart and pending rotates will be retried